### PR TITLE
improve topath perf

### PR DESCRIPTION
**Pull request by AviAvni**

### CODE REVIEW

**Comment by filipecosta90** [comment]
### Automated performance analysis summary

This comment was automatically generated given there is performance data available.

In summary:
- Detected a total of 14 stable tests between versions.
- Detected a total of 2 highly unstable benchmarks.
- Detected a total of 3 improvements above the improvement water line.

You can check a comparison in detail via the [grafana link](https://benchmarksrediscom.grafana.net/d//SH9_rQYGz/?var-branch=master&var-branch=improve-topath-perf)

### Comparison between master and improve-topath-perf.

Time Period from 30 days ago. (environment used: oss-standalone)

|                               Test Case                               |Baseline master (median obs. +- std.dev)|Comparison improve-topath-perf (median obs. +- std.dev)|% change (higher-better)|            Note             |
|-----------------------------------------------------------------------|----------------------------------------|-------------------------------------------------------|------------------------|-----------------------------|
|ENTITY_COUNT                                                           | 33324 +- 0.0%  (7 datapoints)          | 33324 +- nan%  (1 datapoints)                         |-0.0%                   |-- no change --              |
|EXPLICIT-EDGE-DELETION                                                 | 100 +- 0.0%  (7 datapoints)            | 100 +- nan%  (1 datapoints)                           |0.0%                    |-- no change --              |
|GRAPH500-SCALE_18-EF_16-1_HOP                                          | 20000 +- 22.0% UNSTABLE (7 datapoints) | 19999 +- nan%  (1 datapoints)                         |-0.0%                   |UNSTABLE (very high variance)|
|GRAPH500-SCALE_18-EF_16-1_HOP_MIXED_READ_65perc_WRITE_25perc_DEL_10perc| 50 +- 2.6%  (7 datapoints)             | 53 +- nan%  (1 datapoints)                            |5.3%                    |IMPROVEMENT                  |
|GRAPH500-SCALE_18-EF_16-1_HOP_MIXED_READ_75perc_WRITE_25perc           | 59 +- 0.0%  (7 datapoints)             | 59 +- nan%  (1 datapoints)                            |-0.0%                   |-- no change --              |
|GRAPH500-SCALE_18-EF_16-2_HOP                                          | 16665 +- 0.0%  (7 datapoints)          | 16664 +- nan%  (1 datapoints)                         |-0.0%                   |-- no change --              |
|GRAPH500-SCALE_18-EF_16-2_HOP_MIXED_READ_75perc_WRITE_25perc           | 59 +- 0.0%  (7 datapoints)             | 59 +- nan%  (1 datapoints)                            |-0.0%                   |-- no change --              |
|GRAPH500-SCALE_18-EF_16-3_HOP                                          | 4348 +- 0.0%  (7 datapoints)           | 4348 +- nan%  (1 datapoints)                          |-0.0%                   |-- no change --              |
|GRAPH500-SCALE_18-EF_16-3_HOP_MIXED_READ_75perc_WRITE_25perc           | 56 +- 0.0%  (7 datapoints)             | 56 +- nan%  (1 datapoints)                            |-0.0%                   |-- no change --              |
|IMPLICIT-EDGE-DELETION                                                 | 17 +- 0.0%  (7 datapoints)             | 17 +- nan%  (1 datapoints)                            |0.0%                    |-- no change --              |
|INSPECT-EDGES                                                          | 114 +- 13.4% UNSTABLE (7 datapoints)   | 119 +- nan%  (1 datapoints)                           |4.8%                    |UNSTABLE (very high variance)|
|MIX_READ_WRITE                                                         | 11111 +- 3.8%  (7 datapoints)          | 11111 +- nan%  (1 datapoints)                         |-0.0%                   |-- no change --              |
|NODE-INDEX-LOOKUP                                                      | 33 +- 0.0%  (7 datapoints)             | 33 +- nan%  (1 datapoints)                            |-0.0%                   |-- no change --              |
|SORT_ENTITIES                                                          | 100 +- 0.0%  (7 datapoints)            | 100 +- nan%  (1 datapoints)                           |-0.0%                   |-- no change --              |
|UPDATE-BASELINE                                                        | 24999 +- 0.0%  (7 datapoints)          | 24998 +- nan%  (1 datapoints)                         |-0.0%                   |-- no change --              |
|VARIABLE-LENGTH-FILTER                                                 | 33326 +- 0.0%  (7 datapoints)          | 33325 +- nan%  (1 datapoints)                         |-0.0%                   |-- no change --              |
|VARIABLE_LENGTH_EXPAND_INTO                                            | 100 +- 0.0%  (7 datapoints)            | 100 +- nan%  (1 datapoints)                           |0.0%                    |-- no change --              |
|allShortestPaths-10hop-100Kpaths                                       | 16 +- 2.1%  (7 datapoints)             | 25 +- nan%  (1 datapoints)                            |58.3%                   |IMPROVEMENT                  |
|allShortestPaths-4hop-10Kpaths                                         | 217 +- 3.2%  (7 datapoints)            | 294 +- nan%  (1 datapoints)                           |35.3%                   |IMPROVEMENT                  |

File: src/arithmetic/path_funcs/path_funcs.c [hunk 1]
```diff
@@ -26,41 +26,67 @@ SIValue AR_TOPATH(SIValue *argv, int argc, void *private_data) {
...
-			SIValue_Free(path);
+			array_free(arr);
 			return SI_NullVal();
 		}
 
 		if(i % 2 == 0) {
+			path_elements++;
 			// Nodes are in even position.
-			SIPathBuilder_AppendNode(path, element);
+			array_append(arr, element);
```
Comment (swilly22) [inline src/arithmetic/path_funcs/path_funcs.c:43]: There's no validation that `element` is of type `Node`

File: src/arithmetic/path_funcs/path_funcs.c [hunk 2]
```diff
@@ -26,41 +26,67 @@ SIValue AR_TOPATH(SIValue *argv, int argc, void *private_data) {
...
+				array_append(arr, element);
+			} else { // If element is not an edge, it is a path.				
 				/* Path with 0 edges should not be appended. Their source and destination nodes are the same,
 				 * and the source node already appended.
-				 * The build should continue to the next edge/path value. Consider the following query:
-				 * "MATCH p=(a:L1)-[*0..]->(b:L1)-[]->(c:L2)" for the graph in the form of (:L1)-[]->(:L2). The path build should
-				 * return a path with with the relevant entities.
 				 */
-				if(SIPath_Length(element) == 0) {
+				size_t len = SIPath_Length(element);
```
Comment (swilly22) [inline src/arithmetic/path_funcs/path_funcs.c:54]: Are we assuming `element` is of type `Path` ?
Should we validate this assumption ?

File: src/arithmetic/path_funcs/path_funcs.c [hunk 3]
```diff
@@ -26,41 +26,67 @@ SIValue AR_TOPATH(SIValue *argv, int argc, void *private_data) {
...
+				if(len == 0) {
 					i++;
 					continue;
 				}
+				path_elements += len * 2;
```
Comment (swilly22) [inline src/arithmetic/path_funcs/path_funcs.c:59]: Please add a comment explaining this line

File: src/arithmetic/path_funcs/path_funcs.c [hunk 4]
```diff
@@ -26,41 +26,67 @@ SIValue AR_TOPATH(SIValue *argv, int argc, void *private_data) {
 	uint nelements = cypher_ast_pattern_path_nelements(ast_path);
 	ASSERT(argc == (nelements + 1));
 
-	SIValue path = SIPathBuilder_New(nelements);
+	uint path_elements = 0;
+	SIValue *arr = array_new(SIValue, nelements);
 	for(uint i = 0; i < nelements; i++) {
```
Comment (swilly22) [inline src/arithmetic/path_funcs/path_funcs.c:31]: Add a comment explaining the purpose of this loop

File: src/arithmetic/path_funcs/path_funcs.c [hunk 5]
```diff
@@ -26,41 +26,67 @@ SIValue AR_TOPATH(SIValue *argv, int argc, void *private_data) {
 	uint nelements = cypher_ast_pattern_path_nelements(ast_path);
 	ASSERT(argc == (nelements + 1));
 
-	SIValue path = SIPathBuilder_New(nelements);
+	uint path_elements = 0;
+	SIValue *arr = array_new(SIValue, nelements);
```
Comment (swilly22) [inline src/arithmetic/path_funcs/path_funcs.c:30]: I think we can switch from a heap based array to a stack based one

**Review by swilly22 — CHANGES_REQUESTED** [review]
good improvement, please see comments

File: src/arithmetic/path_funcs/path_funcs.c [hunk 6]
```diff
@@ -26,41 +26,69 @@ SIValue AR_TOPATH(SIValue *argv, int argc, void *private_data) {
 	uint nelements = cypher_ast_pattern_path_nelements(ast_path);
 	ASSERT(argc == (nelements + 1));
 
-	SIValue path = SIPathBuilder_New(nelements);
+	uint n = 0;
+	uint path_elements = 0;
+	SIValue arr[nelements];
```
[the remark this answers is not in the document]
Comment (swilly22) [reply src/arithmetic/path_funcs/path_funcs.c:31]: good point!

**Pull request merged**